src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs

```diff
@@ -551,45 +552,91 @@ public override bool HasCustomAttribute(string attributeNamespace, string attrib
                 attributeNamespace, attributeName).IsNil;
         }
 
-        public override int GetInlineArrayLength()
+        public override ClassLayoutMetadata GetClassLayout()
         {
-            Debug.Assert(this.IsInlineArray);
+            TypeLayout layout = _typeDefinition.GetLayout();
+            int inlineArrayLength = 0;
 
-            var attr = MetadataReader.GetCustomAttribute(MetadataReader.GetCustomAttributeHandle(_typeDefinition.GetCustomAttributes(),
-                "System.Runtime.CompilerServices", "InlineArrayAttribute"));
+            if (IsInlineArray)
+            {
+                var attr = MetadataReader.GetCustomAttribute(MetadataReader.GetCustomAttributeHandle(_typeDefinition.GetCustomAttributes(),
+                    "System.Runtime.CompilerServices", "InlineArrayAttribute"));
 
-            var value = attr.DecodeValue(new CustomAttributeTypeProvider(_module)).FixedArguments[0].Value;
+                var value = attr.DecodeValue(new CustomAttributeTypeProvider(_module)).FixedArguments[0].Value;
 
-            return value is int intValue ? intValue : 0;
-        }
+                inlineArrayLength = value is int intValue ? intValue : 0;
+            }
 
-        public override ClassLayoutMetadata GetClassLayout()
-        {
-            TypeLayout layout = _typeDefinition.GetLayout();
+            MetadataLayoutKind layoutKind = MetadataLayoutKind.Auto;
+            if (IsSequentialLayout)
```

Comment:
We can switch directly over `_typeDefinition.Attributes` - no need to go through virtual methods.